File: 51-n-queens/n-queens.go
import (
    "strings"
)

func isValid(placement [][]string, rowIdx, colIdx int) bool{
    // check row 
    for j := range colIdx{
        if placement[rowIdx][j] == "Q"{
            return false
        }
    }

    // check upper diag
    i, j := rowIdx - 1, colIdx - 1
    for i >= 0 && j >= 0{
        if placement[i][j] == "Q"{
            return false
        }
        i--
        j--
    }

    // check lower diag
    i, j = rowIdx + 1, colIdx - 1
    for i < len(placement) && j >= 0{
        if placement[i][j] == "Q"{
            return false
        }
        i++
        j--
    }

    return true
}

func placeQueens(colIdx int, placement [][]string, allPlacements *[][]string) {
    if colIdx == len(placement){
        // add current placement to allPlacements
        currPlacement := make([]string, 0, len(placement))
        for i := range len(placement){
            var builder strings.Builder 
            for j:= range len(placement){
                builder.WriteString(placement[i][j])    
            }
            currPlacement = append(currPlacement, builder.String())
        }
        *allPlacements = append(*allPlacements, currPlacement)
        return
    }

    // choose where to place the Queen
    for rowIdx := range len(placement){
        if isValid(placement, rowIdx, colIdx){
            placement[rowIdx][colIdx] = "Q"
            placeQueens(colIdx + 1, placement, allPlacements)
            placement[rowIdx][colIdx] = "."
        }
    }

}


func solveNQueens(n int) [][]string {
    allPlacements := make([][]string, 0)
    placement := make([][]string, n)
    for i := 0; i < n; i++ {
        placement[i] = make([]string, n)
        for j := 0; j < n; j++ {
            placement[i][j] = "."
        }
    }
    placeQueens(0, placement, &allPlacements)

    return allPlacements
}

File: 51-n-queens/n-queens.py
class Solution:
    def solveNQueens(self, n: int) -> List[List[str]]:
        all_combs = []
        curr_comb = [["."] * n for _ in range(n)]

        def valid_placement(row_idx: int, col_idx: int) -> bool:
            if any(curr_comb[i][col_idx] == "Q" for i in range(row_idx)):
                return False
            
            if any(curr_comb[row_idx][j] == "Q" for j in range(col_idx)):
                return False
            
            i, j = row_idx - 1, col_idx - 1
            while i >= 0 and j >= 0:
                if curr_comb[i][j] == "Q":
                    return False
                i -= 1
                j -= 1

            i, j = row_idx + 1, col_idx - 1
            while i < n and j >= 0:
                if curr_comb[i][j] == "Q":
                    return False
                i += 1
                j -= 1
            
            return True
        
        def find_placements(col_idx: int) -> None:
            # Out of Bounds
            if col_idx >= n:
                copy = curr_comb
                all_combs.append(["".join(row[:]) for row in copy])
                return 
            
            # place queens column wise
            for row_idx in range(n):
                if valid_placement(row_idx, col_idx):
                    curr_comb[row_idx][col_idx] = "Q"
                    find_placements(col_idx + 1)
                    curr_comb[row_idx][col_idx] = "."
            return
        
        find_placements(0)

        return all_combs


